src/SimpleJson/SimpleJson.cs

```diff
@@ -1412,6 +1415,21 @@ public virtual object DeserializeObject(object value, Type type)
                             dict.Add(kvp.Key, DeserializeObject(kvp.Value, valueType));
 
                         obj = dict;
+
+#if SIMPLE_JSON_READONLY_COLLECTIONS
+                        // Wrap dictionary in a ReadOnlyDictionary<,>
+                        var genericTypeDefinition = type.GetGenericTypeDefinition();
+                        if (genericTypeDefinition == typeof(IReadOnlyDictionary<,>) ||
+                            genericTypeDefinition == typeof(ReadOnlyDictionary<,>))
+                        {
+                            var ctorType = typeof(IDictionary<,>).MakeGenericType(keyType, valueType);
+                            var genericReadonlyType = typeof(ReadOnlyDictionary<,>).MakeGenericType(keyType, valueType);
+                            var ctor = ReflectionUtils.GetContructor(genericReadonlyType, new Type[] { ctorType });

```

Comment:
So I was curious why this wasn't a problem with the `ReadOnlyCollection<T>` and `ReadOnlyList<T>`. It turns out that `List<T>` implements `IReadOnlyList<T>` and `IReadOnlyCollection<T>`.

So if we change the dictionary code to create a regular `Dictionary<K,V>` and return that, it _should_ just work.

The one thing I don't like about that is a caller can mutate the "readonly" collections and dictionary by casting it to the concrete type.

So there's several options:

1. Do nothing. This is good enough. After all, it doesn't regress any existing situations.
2. Change the code to return a `Dictionary<K,V>` which fulfills the `IReadOnlyDictionary<K,V>` contract. What this means is we don't add support for deserializing a `ReadOnlyDictionary<K, V>`.
3. Add a new `ReadOnlyCollectionConstructorCache` that supports all the readonly collections. This would allow us to add support for the concrete readonly collections.
4. Change `ConstructorCache` to understand readonly collections.

@prabirshrestha which of these options do you most prefer. I'm leaning away from 4, but 1 - 3 seem acceptable to me :smile: and I'm happy to do the work.